src/Features/CSharp/Portable/Completion/CompletionProviders/CrefCompletionProvider.cs

```diff
@@ -35,178 +43,267 @@ public override bool IsTriggerCharacter(SourceText text, int characterPosition,
             return CompletionUtilities.IsTriggerCharacter(text, characterPosition, options);
         }
 
-        protected override Task<bool> IsExclusiveAsync(Document document, int position, CompletionTriggerInfo triggerInfo, CancellationToken cancellationToken)
+        public override async Task ProduceCompletionListAsync(CompletionListContext context)
         {
-            return SpecializedTasks.True;
-        }
+            var document = context.Document;
+            var position = context.Position;
+            var options = context.Options;
+            var cancellationToken = context.CancellationToken;
 
-        protected override async Task<IEnumerable<CompletionItem>> GetItemsWorkerAsync(Document document, int position, CompletionTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken)
-        {
             var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
             if (!tree.IsEntirelyWithinCrefSyntax(position, cancellationToken))
             {
-                return null;
+                return;
             }
 
-            var token = tree.FindTokenOnLeftOfPosition(position, cancellationToken);
-            token = token.GetPreviousTokenIfTouchingWord(position);
+            var token = tree.FindTokenOnLeftOfPosition(position, cancellationToken)
```

Comment:
I dislike this pattern too. Although, amusingly, it's *your* pattern from way back when C# keyword completion was written, which is why it's everywhere. VB has a GetTargetToken method which I like quite a bit and the CSharpSyntaxContent class used by keyword recommendation and symbol completion has a TargetToken property that already has this value. When I clean up more of the providers, I'll create a GetTargetToken method to use throughout.

I'm kind of surprised it works without diving into trivia as well. Good point. I'll explore this a bit and see what's up.